xarray/core/variable.py

```diff
@@ -893,16 +893,27 @@ def reduce(self, func, dim=None, axis=None, keep_attrs=False,
         if dim is not None and axis is not None:
             raise ValueError("cannot supply both 'axis' and 'dim' arguments")
 
+        if 'cum' in func.__name__:
```

Comment:
I still don't like this approach. It feels very fragile.

A slightly saner approach would another function attribute like `numeric_only` that we use in `ops.py`. Then this check could be: `keep_dims = getattr(func, 'keep_dims', False)`.